### ISSUE TITLE
SonarCloud

### ISSUE THREAD

**Issue by Rapster** [primefaces-extensions/primefaces-extensions.github.com#544]
I've been using Sonar Lint for a while now to improve code quality and became a big fan of it. I was thinking of integrate SonarCloud to PFE _(would be really nice if PF would do the same)_ I can tell you there is a lot to improve (at least in PF from what I saw). @melloware @tandraschko  what do you think?

**Comment by melloware** [comment]
@Rapster let’s definitely do this on PFE and this then we can maybe propose if It to PF after we get it working on PFE

I am a huge fan of code quality tools!

**Comment by Rapster** [comment]
Great I'm on it! :wink:

**Comment by Rapster** [comment]
I'm not admin of the repo so can't activate Travis build :/ @tandraschko or @ova2 do you think you could activate the build please?

**Comment by tandraschko** [comment]
melloware should be able to do it, too.
basically sonar is a good thing, however i don't think we can improve our code much by doing this ;)
In PF there are so many open things, this is one with a low prio.

**Comment by Rapster** [comment]
> basically sonar is a good thing, however i don't think we can improve our code much by doing this ;)

I do not agree, code quality is a big plus (which by definition is a nice improvement :wink:), from there comes a better readability. Tools like Sonar help us to prevent bugs happening. Code from PFE is quite clean so I can't say much, it's another story in PF.

Unfortunately, SonarCloud doesn't assess Javascript code but SonarLint does. I have to dig a bit deeper to see if that's possible

**Comment by melloware** [comment]
I will see if I can activate the travis build.  I will let you know if I run into any issues.

**Comment by melloware** [comment]
OK I got Travis set up but and added the GitHub badge for the build status!

However its failing build because the Master POM 6.3-SNAPSHOT is not in Sonatype Snapshots so I need to push a SNAPSHOT release of the Master Pom to Sonatype.

**Comment by tandraschko** [comment]
I would the master-pom and PFE in one build
same is done in PF for the maven-plugin and PF

**Comment by melloware** [comment]
Let me look how it is done there. Thanks for the tip!

**Comment by melloware** [comment]
OK that worked!  Thanks @tandraschko the build now works except for the last step it needs a Sonar username and password.

```
[INFO] User cache: /home/travis/.sonar/cache
[INFO] SonarQube version: 7.1.0
[INFO] Default locale: "en_US", source code encoding: "iso-8859-1"
[INFO] Publish mode
[INFO] Load global settings
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 36.885 s
[INFO] Finished at: 2018-03-25T17:00:20Z
[INFO] Final Memory: 55M/983M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.4.0.905:sonar (default-cli) on project primefaces-extensions: Not authorized. Please check the properties sonar.login and sonar.password. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by Rapster** [comment]
Alright I'm having a look

**Comment by Rapster** [comment]
@melloware Can you give it a try? I have no idea what's wrong with the token... Here is the procedure https://about.sonarcloud.io/get-started/

**Comment by melloware** [comment]
Sent you an email.  BTW these instructions are terrrible.  Too many "keys" etc.

**Comment by Rapster** [comment]
I gave another try with another tool to encrypt the token, but still not working... There must be something missing I don't know... :@

**Comment by melloware** [comment]
@Rapster I think your organization is incorrect.  Are you sure your "key" is primefaces-extensions.

![image](https://user-images.githubusercontent.com/4399574/37903530-473e0bfc-30c6-11e8-8b61-d7c5313f128d.png)

In my screenshot by Organization Name is "primefaces-extensions" but the KEY is "pr1m3fac3sext3ns1ons"  they want the organization key not the name.

**Comment by Rapster** [comment]
No it's correct, yours is correct too... 

![image](https://user-images.githubusercontent.com/7521311/37904293-9d5b6b94-30fb-11e8-822d-5a1bbdaa1f62.png)

**Comment by melloware** [comment]
OK then I am stumped.   is there any way to ask Sonar for help on why we are getting this issue?   For now I will remove "sonar:sonar" so the build passes.

**Comment by Rapster** [comment]
@melloware I sent a request to the SonarCloud team this morning and answered this:
> Hi Sébastien,

I looked at your code and the last build that failed. Everything looks good into the code. Since you managed to run an analysis locally, I suspect that the token was not correctly encrypted.

To debug this, you can do the following:
Add the following line as the first entry in the "script" section of the Travis YML file:
echo $SONARQUBE_SCANNER_PARAMS
Once the next build is finished, go into Travis CI logs to check what is displayed in this environment variable
=> You should look for the value of the "sonar.login" element and compare this value to your user token
Important: once this test is finished, remove the "echo" line and delete the logs of the Travis build because your token is publicly visible in those logs

I'll have a look later :wink:

**Comment by melloware** [comment]
Cool do you mind doing the following...

1. Change the organization key to "pr1m3fac3sext3ns1ons"

2. Can you encrypt "pr1m3fac3sext3ns1ons" and make that the secure token?   

I don't have RubyGems installed and its a PITA I tried getting it working through Cygwin but its failing.

**Comment by Rapster** [comment]
That's not the key that needs to be encrypted but the token you generate with SonarCloud

**Comment by melloware** [comment]
That is my token..

![image](https://user-images.githubusercontent.com/4399574/37930268-41bf0d80-3110-11e8-8827-9f028355aecb.png)

**Comment by Rapster** [comment]
This is the name of your token not the token itself, SonarCloud should have generate a token, careful it does not keep it (so better to keep it somewhere safe)

![image](https://user-images.githubusercontent.com/7521311/37930461-179c9f76-3143-11e8-9768-d9bf0fa4d059.png)

**Comment by melloware** [comment]
Ahh I see I just generated a new one...

5d7ffc724e6767da4716c2ab5c525802513869b7

**Comment by Rapster** [comment]
I updated .travis.yml with your credantials

**Comment by melloware** [comment]
ok post deleted

**Comment by melloware** [comment]
Damn it failed with "You're only authorized to execute a local (preview) SonarQube analysis without pushing the results to the SonarQube server. Please contact your SonarQube administrator"

**Comment by Rapster** [comment]
I'll write to Sonar's team

**Comment by melloware** [comment]
OK cool.  Let me know if I need to do anything.

**Comment by Rapster** [comment]
> Hi Sébastien,

> From what I see on that build, you managed to run successfully an analysis. Great.
> 
> Now, I see in the next commit that you changed the organization to "pr1m3fac3sext3ns1ons". In this new organization, you (Sébastien) are not part of the members, so you don't have the rights to push an analysis. So if you encrypted your own token, by definition you won't have the right to push an analysis on it. Only the "melloware" user can - since this user is the only member and admin of that organization.

Strange since I used your token...

**Comment by melloware** [comment]
Right did you reply that we work together and you encrypted the token I generated from my account?

**Comment by Rapster** [comment]
Nop but I hmay have a clue of what's going on.. I wonder if it does use my git credentials when I encrypted your token. Can you generate your travis token with this please https://travis-encrypt.github.io/?

**Comment by melloware** [comment]
Yep I just encrypted it and its failing with the same error.

Here is what that travis encyrpt tools spit out

```
Requesting Travis public key from https://api.travis-ci.org/repos/primefaces-extensions/core/key

Public key of primefaces-extensions/core:

-----BEGIN RSA PUBLIC KEY----- MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCkszjAD2EQbbSNksPa5whzobh9 08rAjIwYFpQTGjt5fqL4xThoUIpjQbHoMsdAb+yRo4olHPf5oFy2CJX2ou9JSqOt yt0YWzdZ75Wls0poibRxchIvP9sA26VsvrDo2dLNsQc51Xj26C0GNMM/a6PZ21JY 3eTpMBfi773KOv6wMQIDAQAB -----END RSA PUBLIC KEY-----

Encrypted data:

QntGkT4MuZPXBEvVFuLJ+Y1X1JkmgIDbqhf/Qh9jRrpM3d7u0QKZ7WtuEf/sQ7avkD7FE4642J0nPrdL/uLi7ygE+kbeNCmcom2zZgUSTlUi5SX0EJ8HUHPxAPG1gCtUnrfH1DRwzerHV3sT45S1elR/ailBPQCM2EaEzwVWi2g=
```

**Comment by melloware** [comment]
I opened my own ticket with Sonar so I can communicate with them directly to resolve this.

**Comment by Rapster** [comment]
Yeah sounds like a good idea ;-)

**Comment by Rapster** [comment]
So, since you removed the echo it does not work?

**Comment by melloware** [comment]
@Rapster You have to delete your SonarCloud Project.  Here is what they said...

> Hi,
> 
> I took a closer look, and it turns out that that the "PrimeFaces Extensions" project is already analysed in the "primefaces-extensions" organisation => https://sonarcloud.io/dashboard?id=org.primefaces.extensions%3Aprimefaces-extensions
> This is what the analysis is failing when you try to analyse it in another organisation.
> 
> From there, you have 2 options:
> either you delete this project from the first organisation
> or you change the project key when you analyse it in the 2nd organisation
> I have the feeling that it would be best to go with option 1 - since I don't see any reason for you to have the same project in 2 different organisation.

**Comment by Rapster** [comment]
Alright it's done! :wink:

**Comment by Rapster** [comment]
And it's working at last! :+1:

**Comment by melloware** [comment]
YEAH!!!!

**Comment by melloware** [comment]
I just added the Sonar Quality Gate badge to the Core README.MD.
